cpp/src/arrow/extension/fixed_shape_tensor.cc

```diff
@@ -0,0 +1,267 @@
+      break;
+    }
+    case Type::UINT32: {
+      value_array = std::make_shared<UInt32Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT32: {
+      value_array = std::make_shared<Int32Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::UINT64: {
+      value_array = std::make_shared<Int64Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT64: {
+      value_array = std::make_shared<Int64Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::HALF_FLOAT: {
+      value_array = std::make_shared<HalfFloatArray>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::FLOAT: {
+      value_array = std::make_shared<FloatArray>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::DOUBLE: {
+      value_array = std::make_shared<DoubleArray>(tensor->size(), tensor->data());
+      break;
+    }
+    default: {
+      return Status::NotImplemented("Unsupported tensor type: ",
+                                    tensor->type()->ToString());
+    }
+  }
+  arr = std::make_shared<FixedSizeListArray>(ext_type->storage_type(), tensor->shape()[0],
+                                             value_array);
+  auto ext_data = arr->data();
+  ext_data->type = ext_type;
+  return MakeArray(ext_data);
+}
+
+Result<std::shared_ptr<Tensor>> FixedShapeTensorType::ToTensor(
+    std::shared_ptr<Array> arr) const {
+  // To convert an array of n dimensional tensors to a n+1 dimensional tensor we
+  // interpret the array's length as the first dimension the new tensor. Further, we
+  // define n+1 dimensional tensor's strides by front appending a new stride to the n
+  // dimensional tensor's strides.
+
+  ARROW_DCHECK_EQ(arr->null_count(), 0) << "Null values not supported in tensors.";
+  auto ext_arr = internal::checked_pointer_cast<FixedSizeListArray>(
+      internal::checked_pointer_cast<ExtensionArray>(arr)->storage());
+
+  std::vector<int64_t> shape = shape_;
+  shape.insert(shape.begin(), 1, arr->length());
+
+  std::vector<int64_t> tensor_strides = strides();
+  tensor_strides.insert(tensor_strides.begin(), 1, arr->length() * tensor_strides[0]);
+
+  std::shared_ptr<Buffer> buffer = ext_arr->values()->data()->buffers[1];
+  return *Tensor::Make(ext_arr->value_type(), buffer, shape, tensor_strides, dim_names());
+}
+
+std::shared_ptr<FixedShapeTensorType> fixed_shape_tensor(
+    const std::shared_ptr<DataType>& value_type, const std::vector<int64_t>& shape,
+    const std::vector<int64_t>& permutation, const std::vector<std::string>& dim_names) {
+  ARROW_CHECK(is_tensor_supported(value_type->id()));
+
+  if (!permutation.empty()) {
+    ARROW_CHECK_EQ(shape.size(), permutation.size())
+        << "permutation.size() == " << permutation.size()
+        << " must be empty or have the same length as shape.size() " << shape.size();
+  }
+  if (!dim_names.empty()) {
+    ARROW_CHECK_EQ(shape.size(), dim_names.size())
```

Comment:
Also check that all elements are strings?

Reply:
And are those errors already tested?

Reply:
I'm not sure how to assert for these in tests. `ASSERT_DEATH` would work but it seems it's not advised.
We also can't change the return type of the function so we can't return `Status` or `Result`.

Reply:
@benibus any recommendations here?

Reply:
Maybe add a static `Make` function to `FixedShapeTensorType` that does any pre-validation and returns a `Result`, which you can use in the tests instead. Then in `fixed_shape_tensor` you can do something like
```
auto maybe_type = FixedShapeTensorType::Make(value_type, shape, permutation, dim_names);
ARROW_CHECK_OK(maybe_type);
return maybe_type.MoveValueUnsafe();
```

Now that I'm thinking about it, it may not be appropriate to do extensive validation/assertions in the lowercase factory functions, since I think it breaks precedence. Typically, those functions will just call `std::make_shared<WhateverType>(...)` with the given arguments and run the risk of the output being invalid. Instead, it might be best to do something like this
```
std::vector<int64_t> strides;
DCHECK_OK(ComputeStrides(value_type, shape, permutation, &strides));
return std::make_shared<FixedShapeTensorType>(value_type, shape, permutation, strides);
```
So you would pre-compute the stride vector and pass it into the constructor. Then, in `FixedSizeTensorType::Make` you could do the more extensive checks in addition to propagating any errors from `ComputeStrides`.

Reply:
For a similar example of what I mean, see https://github.com/apache/arrow/blob/a3cd962b8bc5c49f33824202dff382af13cfa873/cpp/src/arrow/type.h#L1605-L1611 and https://github.com/apache/arrow/blob/a3cd962b8bc5c49f33824202dff382af13cfa873/cpp/src/arrow/type.cc#L2541-L2545

Reply:
Yeah, that makes a lot of sense, changed. `ComputeStrides` is now called lazily when calling `.strides`.